File: src/slib/network/url_request.cpp
/*
 *   Copyright (c) 2008-2018 SLIBIO <https://github.com/SLIBIO>
 *
 *   Permission is hereby granted, free of charge, to any person obtaining a copy
 *   of this software and associated documentation files (the "Software"), to deal
 *   in the Software without restriction, including without limitation the rights
 *   to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 *   copies of the Software, and to permit persons to whom the Software is
 *   furnished to do so, subject to the following conditions:
 *
 *   The above copyright notice and this permission notice shall be included in
 *   all copies or substantial portions of the Software.
 *
 *   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 *   IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 *   FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 *   AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 *   LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 *   OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 *   THE SOFTWARE.
 */

#include "slib/network/url_request.h"

#include "slib/network/url.h"
#include "slib/core/json.h"
#include "slib/core/thread_pool.h"
#include "slib/core/safe_static.h"
#include "slib/core/event.h"
#include "slib/core/system.h"

namespace slib
{

	namespace priv
	{
		namespace url_request
		{
			typedef HashMap< UrlRequest*, Ref<UrlRequest> > UrlRequestMap;
			SLIB_SAFE_STATIC_GETTER(UrlRequestMap, getUrlRequestMap)

		}
	}
	
	using namespace priv::url_request;
	
	SLIB_DEFINE_OBJECT(UrlRequest, Object)
	
	UrlRequest::UrlRequest()
	{
		m_sizeBodySent = 0;
		m_sizeContentTotal = 0;
		m_sizeContentReceived = 0;
		m_flagClosed = sl_false;
		m_flagError = sl_false;
		
		m_responseStatus = HttpStatus::Unknown;
		
		m_method = HttpMethod::GET;
		m_flagSelfAlive = sl_false;
		m_flagStoreResponseContent = sl_true;
		m_flagUseBackgroundSession = sl_false;

	}
	
	UrlRequest::~UrlRequest()
	{
	}
	
#define URL_REQUEST UrlRequest
#include "url_request_common.inc"
	
	const String& UrlRequest::getUrl()
	{
		return m_url;
	}
	
	const String& UrlRequest::getDownloadingFilePath()
	{
		return m_downloadFilePath;
	}
	
	HttpMethod UrlRequest::getMethod()
	{
		return m_method;
	}
	
	const Memory& UrlRequest::getRequestBody()
	{
		return m_requestBody;
	}
	
	sl_size UrlRequest::getRequestBodySize()
	{
		return m_requestBody.getSize();
	}
	
	HashMap<String, Variant> UrlRequest::getParameters()
	{
		return m_parameters;
	}
	
	HttpHeaderMap UrlRequest::getRequestHeaders()
	{
		return m_requestHeaders;
	}
	
	Memory UrlRequest::getResponseContent()
	{
		return m_bufResponseContent.merge();
	}
	
	String UrlRequest::getResponseContentAsString()
	{
		return String::fromUtf(m_bufResponseContent.merge());
	}
	
	Json UrlRequest::getResponseContentAsJson()
	{
		Memory mem = m_bufResponseContent.merge();
		String16 s = String16::fromUtf(mem);
		return Json::parseJson(s);
	}
	
	Ref<XmlDocument> UrlRequest::getResponseContentAsXml()
	{
		Memory mem = m_bufResponseContent.merge();
		String16 s = String16::fromUtf(mem);
		return Xml::parseXml(s);
	}
	
	HttpStatus UrlRequest::getResponseStatus()
	{
		return m_responseStatus;
	}
	
	String UrlRequest::getResponseMessage()
	{
		return m_responseMessage;
	}
	
	HttpHeaderMap UrlRequest::getResponseHeaders()
	{
		return m_responseHeaders;
	}
	
	String UrlRequest::getResponseHeader(const String& name)
	{
		return m_responseHeaders.getValue(name, String::null());
	}
	
	sl_bool UrlRequest::isUsingBackgroundSession()
	{
		return m_flagUseBackgroundSession;
	}
	
	sl_bool UrlRequest::isSelfAlive()
	{
		return m_flagSelfAlive;
	}
	
	sl_bool UrlRequest::isStoringResponseContent()
	{
		return m_flagStoreResponseContent;
	}
	
	sl_uint64 UrlRequest::getSentRequestBodySize()
	{
		return m_sizeBodySent;
	}
	
	sl_uint64 UrlRequest::getResponseContentSize()
	{
		return m_sizeContentReceived;
	}
	
	sl_uint64 UrlRequest::getExpectedResponseContentSize()
	{
		return m_sizeContentTotal;
	}
	
	void UrlRequest::cancel()
	{
		if (m_flagClosed) {
			return;
		}
		ObjectLocker lock(this);
		if (m_flagClosed) {
			return;
		}
		m_flagClosed = sl_true;
		_removeFromMap();
		_cancel();
	}
	
	sl_bool UrlRequest::isError()
	{
		if (m_responseStatus >= HttpStatus::BadRequest) {
			return sl_true;
		}
		return m_flagError;
	}

	String UrlRequest::getLastErrorMessage()
	{
		return m_lastErrorMessage;
	}
	
	sl_bool UrlRequest::isClosed()
	{
		return m_flagClosed;
	}
	
	void UrlRequest::_sendSync()
	{
		Ref<Event> ev = Event::create();
		if (ev.isNotNull()) {
			m_eventSync = ev;
			_sendAsync();
			ev->wait();
			return;
		}
		onError();
	}
	
	void UrlRequest::_sendSync_call()
	{
		_sendSync();
	}

	class UrlRequest_AsyncPool
	{
	public:
		Ref<ThreadPool> threadPool;
		
	public:
		UrlRequest_AsyncPool()
		{
			threadPool = ThreadPool::create();
		}
		
	};
	
	SLIB_SAFE_STATIC_GETTER(UrlRequest_AsyncPool, Get_UrlRequestAsyncPool)
	
	void UrlRequest::_sendAsync()
	{
		UrlRequest_AsyncPool* pool = Get_UrlRequestAsyncPool();
		if (pool) {
			if (pool->threadPool->addTask(SLIB_FUNCTION_WEAKREF(UrlRequest, _sendSync_call, this))) {
				return;
			}
		}
		onError();
	}
		
	void UrlRequest::_init(const UrlRequestParam& param, const String& url)
	{
		m_url = url;
		m_downloadFilePath = param.downloadFilePath;
		
		m_method = param.method;
		m_requestBody = param.requestBody;
		m_parameters = param.parameters;
		m_requestHeaders = param.requestHeaders;

		m_onComplete = param.onComplete;
		m_onResponse = param.onResponse;
		m_onReceiveContent = param.onReceiveContent;
		m_onDownloadContent = param.onDownloadContent;
		m_onUploadBody = param.onUploadBody;
		m_dispatcher = param.dispatcher;
		
		m_flagUseBackgroundSession = param.flagUseBackgroundSession;
		m_flagSelfAlive = param.flagSelfAlive && !(param.flagSynchronous);
		m_flagStoreResponseContent = param.flagStoreResponseContent;
		
		m_timeout = param.timeout;
		m_flagAllowInsecureConnection = param.flagAllowInsecureConnection;
		
		if (m_flagSelfAlive) {
			UrlRequestMap* map = getUrlRequestMap();
			if (map) {
				map->put(this, this);
			}
		}

	}
	
	void UrlRequest::_removeFromMap()
	{
		if (m_flagSelfAlive) {
			UrlRequestMap* map = getUrlRequestMap();
			if (map) {
				map->remove(this);
			}
		}
	}
	
	void UrlRequest::_cancel()
	{
	}
	
	void UrlRequest::onComplete()
	{
		if (m_flagClosed) {
			return;
		}
		m_flagClosed = sl_true;
		if (m_onComplete.isNotNull()) {
			if (m_dispatcher.isNotNull()) {
				m_dispatcher->dispatch(SLIB_BIND_REF(void(), UrlRequest, _runCallback, this, m_onComplete));
			} else {
				m_onComplete(this);
			}
		}
		_removeFromMap();
		
		if (m_eventSync.isNotNull()) {
			m_eventSync->set();
		}
	}
	
	void UrlRequest::onError()
	{
		m_flagError = sl_true;
		onComplete();
	}
	
	void UrlRequest::onResponse()
	{
		if (m_flagClosed) {
			return;
		}
		m_sizeContentReceived = 0;
		m_bufResponseContent.clear();
		m_onResponse(this, m_responseStatus);
	}
	
	void UrlRequest::onReceiveContent(const void* data, sl_size len, const Memory& mem)
	{
		if (m_flagClosed) {
			return;
		}
		if (m_flagStoreResponseContent) {
			if (mem.isNotNull()) {
				m_bufResponseContent.add(mem);
			} else {
				m_bufResponseContent.add(Memory::create(data, len));
			}
			m_sizeContentReceived = m_bufResponseContent.getSize();
		} else {
			m_sizeContentReceived += len;
		}
		m_onReceiveContent(this, data, len);
	}
	
	void UrlRequest::onDownloadContent(sl_uint64 size)
	{
		if (m_flagClosed) {
			return;
		}
		m_sizeContentReceived += size;
		m_onDownloadContent(this, size);
	}
	
	void UrlRequest::onUploadBody(sl_uint64 size)
	{
		if (m_flagClosed) {
			return;
		}
		m_onUploadBody(this, size);
	}
	
	void UrlRequest::_runCallback(const Function<void(UrlRequest*)>& callback)
	{
		callback(this);
	}
	
}

File: java/android/slib/platform/android/network/UrlRequest.java
/*
 *   Copyright (c) 2008-2018 SLIBIO <https://github.com/SLIBIO>
 *
 *   Permission is hereby granted, free of charge, to any person obtaining a copy
 *   of this software and associated documentation files (the "Software"), to deal
 *   in the Software without restriction, including without limitation the rights
 *   to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 *   copies of the Software, and to permit persons to whom the Software is
 *   furnished to do so, subject to the following conditions:
 *
 *   The above copyright notice and this permission notice shall be included in
 *   all copies or substantial portions of the Software.
 *
 *   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 *   IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 *   FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 *   AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 *   LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 *   OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 *   THE SOFTWARE.
 */

package slib.platform.android.network;

import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.net.URLConnection;
import java.util.Vector;

import slib.platform.android.Logger;

public class UrlRequest {

	final static int NETWORK_TIMEOUT = 30000;
	final static int PACKET_SIZE = 4096;

	boolean flagOpened = true;
	long instance;

	URLConnection connection = null;
	HttpURLConnection http = null;
	InputStream streamContent = null;
	OutputStream streamDownload = null;

	UrlRequest(long instance) {
		this.instance = instance;
	}

	public static void execute(long instance, String _url, String method, String[] headers, String[] additionalHeaders, byte[] uploadBody, String downloadFilePath) {

		if (downloadFilePath != null && downloadFilePath.length() == 0) {
			downloadFilePath = null;
		}

		UrlRequest request = new UrlRequest(instance);

		try {
			request.onInit();

			OutputStream streamDownload = null;
			if (downloadFilePath != null) {
				streamDownload = new FileOutputStream(downloadFilePath);
			}
			request.streamDownload = streamDownload;

			URL url = new URL(_url);
			URLConnection connection = url.openConnection();
			request.connection = connection;

			HttpURLConnection http = null;
			if (connection instanceof HttpURLConnection) {
				http = (HttpURLConnection) connection;
			}
			request.http = http;

			if (!(request.flagOpened)) {
				return;
			}
			if (http != null) {
				if (uploadBody != null && uploadBody.length > 0) {
					http.setDoOutput(true);
				}
				http.setDoInput(true);
				http.setConnectTimeout(NETWORK_TIMEOUT);
				http.setReadTimeout(NETWORK_TIMEOUT);
				http.setRequestMethod(method);
				if (headers != null) {
					for (int i = 0; i < headers.length - 1; i += 2) {
						http.setRequestProperty(headers[i], headers[i+1]);
					}
				}
				if (additionalHeaders != null) {
					for (int i = 0; i < additionalHeaders.length - 1; i += 2) {
						http.addRequestProperty(additionalHeaders[i], additionalHeaders[i+1]);
					}
				}
			}

			connection.connect();
			if (!(request.flagOpened)) {
				return;
			}

			if (http != null && uploadBody != null && uploadBody.length > 0) {
				OutputStream os = connection.getOutputStream();
				int pos = 0;
				while (pos < uploadBody.length) {
					if (!(request.flagOpened)) {
						return;
					}
					int len = PACKET_SIZE;
					if (pos + len > uploadBody.length) {
						len = uploadBody.length - pos;
					}
					os.write(uploadBody, pos, len);
					request.onUploadBody(len);
					pos += len;
				}
			}
			if (!(request.flagOpened)) {
				return;
			}

			if (http != null) {
				int statusCode = http.getResponseCode();
				String message = http.getResponseMessage();
				Vector<String> responseHeaders = new Vector<String>();
				for (int i = 0; i < 1024; i++) {
					String key = http.getHeaderFieldKey(i);
					String value = http.getHeaderField(i);
					if (key == null && value == null) {
						break;
					}
					responseHeaders.add(key);
					responseHeaders.add(value);
				}
				request.onResponse(statusCode, message, responseHeaders.toArray(new String[] {}));
			}
			if (!(request.flagOpened)) {
				return;
			}

			InputStream streamContent = null;
			try {
				streamContent = connection.getInputStream();
			} catch (FileNotFoundException e) {}

			request.streamContent = streamContent;

			if (streamContent != null) {
				byte[] packet = new byte[PACKET_SIZE];
				while (true) {
					int n = streamContent.read(packet);
					if (n <= 0) {
						break;
					}
					if (downloadFilePath != null) {
						streamDownload.write(packet, 0, n);
						request.onDownloadContent(n);
					} else {
						request.onReceiveContent(packet, n);
					}
					if (!(request.flagOpened)) {
						return;
					}
				}
			}

			request.close();

			request.onComplete();

		} catch (Exception e) {
			request.processError(e);
		}

	}

	public synchronized void close() {
		if (flagOpened) {
			flagOpened = false;
			if (streamDownload != null) {
				try {
					streamDownload.close();
				} catch (Exception e) {
					Logger.exception(e);
				}
				streamDownload = null;
			}
			connection = null;
			if (http != null) {
				try {
					http.disconnect();
				} catch (Exception e) {
					Logger.exception(e);
				}
				http = null;
				streamContent = null;
			} else {
				if (streamContent != null) {
					try {
						streamContent.close();;
					} catch (Exception e) {
						Logger.exception(e);
					}
					streamContent = null;
				}
			}
		}
	}

	void processError(Exception e) {
		close();
		if (e != null) {
			Logger.exception(e);
			onError(e.getLocalizedMessage());
		} else {
			onError("Unknown Error");			
		}
	}

	private native void nativeOnInit(long instance);
	protected void onInit() {
		nativeOnInit(instance);
	}

	private native void nativeOnUploadBody(long instance, int n);
	protected void onUploadBody(int n) {
		nativeOnUploadBody(instance, n);
	}

	private native void nativeOnResponse(long instance, int statusCode, String message, String[] headers);
	protected void onResponse(int statusCode, String message, String[] headers) {
		nativeOnResponse(instance, statusCode, message, headers);
	}

	private native void nativeOnReceiveContent(long instance, byte[] data, int n);
	protected void onReceiveContent(byte[] data, int n) {
		nativeOnReceiveContent(instance, data, n);
	}

	private native void nativeOnDownloadContent(long intance, int n);
	protected void onDownloadContent(int n) {
		nativeOnDownloadContent(instance, n);
	}

	private native void nativeOnComplete(long instance);
	protected void onComplete() {
		nativeOnComplete(instance);
	}

	private native void nativeOnError(long instance, String errorMessage);
	protected void onError(String errorMessage) {
		nativeOnError(instance, errorMessage);
	}

}
